### ISSUE TITLE
Vacancies rewrite

### ISSUE THREAD

**Issue by hoyes** [camdram/camdram#159]
Placeholder ticket to hold ideas for vacancies section rewrite

**Comment by philosophicles** [comment]
From an email conversation: 

Perhaps what we actually need is a more flexible, single Vacancies page, with filtering facility. More like how Amazon or other shopping sites allow you to filter items for sale. 

I.e. the left hand pane that currently has 3 links for auditions, techie roles and show applications, would turn into a search facility. This would include things like: 

* Tickboxes for roles you want to see, 
* date filtering (audition date from/to; actual production date from/to) etc. 
* Role categories (e.g. all designer roles would be 'Design', true techie roles 'Technical', Cast, dancers, musicians might all be 'On-stage' etc.)

This would probably require us to improve the structuring of vacancy data so isn't a short-term solution.

I'm imagining the filter conditions would be HTTP GET parameters so you could share/bookmark a link with your custom filter conditions, and the different Views (RSS, plain text etc) would all respond to the active filter conditions.

**Comment by jamesterjlrb** [comment]
So an idea that I've had floating about for a bit is of creating a master list of roles that camdram recognises. This would likely be necessary for filtering specific roles. This may be problematic for use in shows listings of production teams where there are sometimes some very esoteric roles listed but for vacancies this may work as many of these esoteric roles can be described as something more common and clarified in the vacancy blurb.

**Comment by hoyes** [comment]
#20 is relevant to this. The idea presented there is to have a hierarchy of roles curated by the Camdram admins (e.g. Production -> Technical -> Lighting -> Production Electrican, or Production -> Design -> Costume Designer etc), then you could filter at any level of specific-ness.

In the case of show production team listings, I was thinking you could allowing esoteric roles still, but if it's not something Camdram knows about it you would somehow be required to also select the most specific role it's similar to (e.g. if it's not a "set builder" or "painter" you could just select "techie"). A bit of a UI challenge admittedly - could be an autocomplete thing with an "other" option which pops up some extra fields.

It would be really cool if the same "role" concept could apply across Camdram - it would enable interesting things down the line like suggesting open vacancies people should apply for based on what positions they held on previous shows, or suggest relevant infobase articles based on what shows people have coming up.

**Comment by jamesterjlrb** [comment]
I totally agree with the "select the nearest" concept. But as you say it sounds tricky to implement which is why I held back from implementing it. You'd want to be able to have "grumpy techie minion" visible but have the system know that it was a "stagehand/crew" (for example). Or if people really wanted to be listed as a lighting director the system would equate it to lighting designer.

**Comment by hoyes** [comment]
You could get so far by teaching Camdram lots of aliases (e.g. Production Electrician, Lighting Assistant, Assistant Lighting Designer, ALD could all be equally valid but be recognised as aliases for the same concept).

The same problem exists in the vacancies world really. You might not get advertisements for a "grumpy techie minion" but you might get valid but unusual things like "flame thrower" or "prosthetics expert"  etc

Pragmatically speaking, the answer is probably to do this in stages:
 * Step 1: Design and build the list/hierarchy (this is happen pre v1-switch-off thinking about it)
 * Step 2: Build a new vacancies section around it
 * Step 3: Figure out how to integrate show listings.

**Comment by philosophicles** [comment]
Rather than teach loads of aliases, would it be easier to allow a defined set of roles (as you suggest) based on skills involved, and allow the show admin(s) to enter a custom description against it? E.g. for a big show they might list requirement for 3 master carpenters, one of which will be described as 'Main Set Carpenter', one as 'Ship Builder' and one as 'Revolve Engineer'.  

We wouldn't catalogue the custom descriptions particularly, except perhaps identifying common ones if we wanted to add auto-complete functionality to the description field (merely to encourage consistency where appropriate). 

This is not all that far from what you suggest but perhaps coming at it from an opposite viewpoint. 

@hoyes I very much like the idea of a hierarchy though. Does the hierarchy have to go down to the lowest level of detail? Or could the lowest curated level be still a sub-category, and users could create their own roles as desired within each subcategory? (Like how on, e.g., Gumtree, you have to list items for sale under a subcategory, but you can sell many quite different things within each lowest level category.)

Another option (not mutually exclusive) would be non-hierarchical tagging - again, we could curate the allowed tags (much like Stack Exchange for e.g.) and allow people to apply as many as appropriate. These two ideas could be used together to give two dimensions of filtering, e.g. tagging to indicate desired skills and hierarchy for the general role category?

**Comment by philosophicles** [comment]
Whilst riffing on ideas is good - I'm approaching this as brainstorming - I also think it'd be worth us actively seeking ideas and opinions beyond the dev team at some stage. E.g. emailing round society committees and significant mailing list groups, asking for ideas (but making clear the active developers ultimately make the decisions).

**Comment by hoyes** [comment]
Bumping this for re-discussion in 2018. @CHTJonas and @GKFX any thoughts?

Our vacancies data model hasn't changed at all since v1 so there's a lot that could be done here. There has always also been certain types of roles that haven't been very well catered for (writers and musicians) and we could do a lot better for the UI than 3 long lists.

A starting point might be to try and unify our current audtions, "techieadverts", and "applications" tables into one base concept, maybe with certain sub-types for, say, specific audition dates. There are some odd and unnecessary differences between how things like contact details and deadlines work for the three tables. Maybe at the same time decoupling this concept from shows (see #320).

I wonder whether we should organise some sort of survey of producers asking things like:
 * What sort of information they'd like to include in listings
 * How they current run auditions/interviews and if there are ways Camdram could help with that
 * Anything else Camdram could do to make their lives easier.
Vacancies have always sort of been the "carrot" that encourages producers to bother adding shows to Camdram in the first place so we don't want to screw it up...

**Comment by GKFX** [comment]
The discussion above seems reasonable; a loose hierarchy of roles, with the ability to advertise in multiple categories where appropriate, and then producers continue to be able to name roles as they feel appropriate. It also makes sense to be able to categorise some roles in the supercategories "techie" or "performer" rather than e.g. "carpenter".

Actor roles would probably be less hierarchical, with more multiple-tagging going on. As I don't know 
anything about acting that is definitely a job for the actors' reps and mailing lists!

StackOverflow allows you to filter by the presence or absence of tags and that makes sense. Users could then search for all non-humourous acting or for a list of favourite techie roles.

The long scrolling list of adverts problem is more about display than about the actual information we hold. In the shiny new vacancies listing, each vacancy could be a summary box (event name, roles available, deadline, event date where applicable/known) with the description and contact details accessible on clicking. (That might be on the /show/ page or a new /vacancy/ page.)

**Comment by CHTJonas** [comment]
First thought: this sounds a bit horrendous to manage as a DB migration, or preserving history going back. But I guess if we time this so that it happens during a vacant 'down' time over summer maybe it won't be as bad?

I love @GKFX's idea of having free-form text entry fields for the title, and then tagging the role with appropriate attributes (so Lampie Minion could be tagged with Lighting Design, Lighting, LD, LX...). This would probably work for actors as well (female lead, male ensemble, female treble, male tenor) and shows even (comedy, musical, play, stand-up).

I do think we should definitely do a consultation/focus group/survey of directors/producers/societies/users before we make this change.

**Comment by hoyes** [comment]
If I've learned one thing about making big software changes over the years it's to try and avoid massive "pull the lever and everything changes" kind of moments. It's more difficult and more tedious but ultimately safer to try and break down whatever plan we come up with into a set of smaller steps than can be managed by DB migrations than can be scripted and tested in advance on an offline database.

We probably also need to ensure the https://www.camdram.net/vacancies/techies.txt etc carry on working sensibly throughout all this.

A possible set of migration steps:
 * Move audition information out of the shows table into a separate "acts_audition_advert" table or something, with the acts_auditions linked to acts_audition_advert instead of acts_shows.
 * Merge all three "vacancy ad" concepts into a single table called "acts_adverts" or something, but with each ad still explicitly tagged as "audition", "techie" or "application" based on the form used to create the ad.
 * Break out an advert's individual roles into a separate table and create the role hierarchy and tagging system. Replace the manual advert "tags" with something that automatically determines relevance for the "audition", "techie" or "application" page based on the tags present
 * Redesign the UI to allow more filtering options that just "audition", "techie", "application".
 * Somewhere along the line, make the link from an "advert" to a "show" optional.
